source/extensions/filters/http/grpc_field_extraction/filter.cc

```diff
@@ -0,0 +1,229 @@
+                                    kRcDetailErrorRequestBufferConversion));
+    return HandleDecodeDataStatus(Envoy::Http::FilterDataStatus::StopIterationNoBuffer);
+  }
+
+  if (buffering->empty()) {
+    ENVOY_STREAM_LOG(debug, "not a complete msg", *decoder_callbacks_);
+    // Not a complete message.
+    return HandleDecodeDataStatus(Envoy::Http::FilterDataStatus::StopIterationNoBuffer);
+  }
+
+  // Buffering returns a list of messages.
+  for (size_t msg_idx = 0; msg_idx < buffering->size(); ++msg_idx) {
+    std::unique_ptr<StreamMessage> message_data = std::move(buffering->at(msg_idx));
+
+    // MessageConverter uses a empty StreamMessage to denote the end.
+    if (message_data->size() == -1) {
+      ABSL_DCHECK(end_stream);
+      ABSL_DCHECK(message_data->isFinalMessage());
+      // This is the last one in the vector.
+      ABSL_DCHECK(msg_idx == buffering->size() - 1);
+      // Skip the empty message
+      continue;
+    }
+
+    if (!extraction_done_) {
+      extraction_done_ = true;
+
+      auto result = extractor_->processRequest(*message_data->message());
+      if (!result.ok()) {
+        rejectRequest(result.status().raw_code(), result.status().message(),
+                      generateRcDetails(kRcDetailFilterGrpcFieldExtraction,
+                                        absl::StatusCodeToString(result.status().code()),
+                                        kRcDetailErrorRequestFieldExtractionFailed));
+        return HandleDecodeDataStatus(Envoy::Http::FilterDataStatus::StopIterationNoBuffer);
+      }
+      handleExtractionResult(*result);
+    }
+
+    auto buf_convert_status = request_msg_converter_->convertBackToBuffer(std::move(message_data));
+    // The message_data is not modified, ConvertBack should return OK.
+    RELEASE_ASSERT(buf_convert_status.ok(), "request message convert back should work");
+
+    data.move(*buf_convert_status.value());
+    ENVOY_STREAM_LOG(debug, "decodeData: convert back data size={}", *decoder_callbacks_,
+                     data.length());
+  }
+
+  // Reject the request if  extraction is required but could not
+  // buffer up any messages.
+  if (!extraction_done_) {
+    rejectRequest(Status::WellKnownGrpcStatus::InvalidArgument,
+                  "did not receive enough data to form a message.",
+                  generateRcDetails(kRcDetailFilterGrpcFieldExtraction,
+                                    absl::StatusCodeToString(absl::StatusCode::kInvalidArgument),
+                                    kRcDetailErrorRequestOutOfData));
+    return HandleDecodeDataStatus(Envoy::Http::FilterDataStatus::StopIterationNoBuffer);
+  }
+  return HandleDecodeDataStatus();
+}
+
+void Filter::handleExtractionResult(const ExtractionResult& result) {
+  ABSL_DCHECK(extractor_);
+
+  ProtobufWkt::Struct dest_metadata;
+  for (const auto& req_field : result) {
+    auto* list = (*dest_metadata.mutable_fields())[req_field.path].mutable_list_value();
```

Comment:
Is `[req_field.path]` guaranteed to exist?

Reply:
Yes.

Reply:
How is it guaranteed?